Rebase policy trace to use L3 policy resolve in L4PolicyMap

joestringer opened the pull request:
Recently, label-based policy was reworked to be implemented via the L4PolicyMap, and resolved as part of the same step as L4 policy resolution. However, the policy tracing remained using the old code paths. This PR seeks to rectify this by reworking the policy trace functions to extend the L4 policy trace to appropriately report messages during policy resolution, and to use this function for the policy trace paths.

The resulting code is a little simpler, which is nice. Most of the changes here are minor test fixups.

Here's an example of the updated text output for a trace where there are three rules:
* Allow from `reserved:host` and `any:foo` at L3
* Allow from `reserved:host` and `any:baz` on port 80
* Require traffic ingressing to `any:bar` to contain the labels `any:baz`.

It's disallowed, because the trace doesn't specify L4 port but the rule only allows port 80.
```
Tracing From: [any:baz] => To: [any:bar] Ports: [0/ANY]

Resolving ingress policy for [any:bar]                                          
* Rule {"matchLabels":{"any:bar":""}}: selected                                 
    Enforcing requirements [{Key:any.baz Operator:In Values:[]}]                
    Allows from labels {"matchLabels":{"reserved:host":""},"matchExpressions":[{"key":"any:baz","operator":"In","values":[""]}]}
    Allows from labels {"matchLabels":{"any:foo":""},"matchExpressions":[{"key":"any:baz","operator":"In","values":[""]}]}
      No label match for [any:baz]                                              
* Rule {"matchLabels":{"any:bar":""}}: selected                                 
    Enforcing requirements [{Key:any.baz Operator:In Values:[]}]                
    Allows from labels {"matchLabels":{"reserved:host":""},"matchExpressions":[{"key":"any:baz","operator":"In","values":[""]}]}
    Allows from labels {"matchLabels":{"any:baz":""},"matchExpressions":[{"key":"any:baz","operator":"In","values":[""]}]}
      Found all required labels                                                 
      Allows port [{80 ANY}]                                                    
        No port match found                                                     
* Rule {"matchLabels":{"any:bar":""}}: selected                                 
    Enforcing requirements [{Key:any.baz Operator:In Values:[]}]                
      Found all required labels                                                 
3/3 rules selected                                                              
Found no allow rule                                                             
Ingress verdict: denied
```

TODO:
* [x] Validate tests pass
* [x] Update documentation examples
  * Prepared additional commit, can open another PR or update this one

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7871)
<!-- Reviewable:end -->

joestringer committed:
policy: Drop usage of deniedIdentities in testing code

This testing code doesn't test requirements anyway, drop the plumbing.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
policy: Add logging helper to SearchContext

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
policy: Support L3 tracing of L4PolicyMap

Recent commits have integrated L3 policies into L4PolicyMap, now add
support for performing policy trace on it for labels-only policy.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
policy/api: Add helper for PortProtocol supersets

Add a helper which determines whether a received PortProtocol are equal
or a superset of another PortProtocol.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
policy: Rework ingress policy trace to L4PolicyMap

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
policy: Improve debuggability of test case

By reworking this test case to save a buffer of the trace, it's easier
to figure out what goes wrong when something goes wrong.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
policy: Rework egress policy trace to L4PolicyMap

Signed-off-by: Joe Stringer <joe@cilium.io>

ianvernon commented on pkg/policy/rule.go:
```diff
@@ -179,19 +180,65 @@ func mergeIngressPortProto(ctx *SearchContext, endpoints []api.EndpointSelector,
 }
 
+func traceL3(ctx *SearchContext, peerEndpoints api.EndpointSelectorSlice, direction string) {
+	var result bytes.Buffer
```
This code should only be ran if `ctx.Trace != TRACE_DISABLED`, since otherwise it's a no-op for policy computation and we'd be creating these buffers and doing string writing without using the output.

joestringer replied:
Good point. I'm inadvertently achieving this today because only the policy trace cases specify both the source and destination (well, apart from some deniedIdentities logic which @jrajahalme is working on removing right now), but this should be more explicit.

ianvernon commented on pkg/policy/rule.go:
```diff
@@ -309,6 +360,12 @@ func (r *rule) resolveIngressPolicy(ctx *SearchContext, state *traceState, resul
+			// If the rule only consists of a FromRequires statement,
+			// provide an empty selector so that it will be updated
+			// in the next step.
+			if len(ruleCopy.FromEndpoints) == 0 {
```
I still don't understand why this is necessary - requirements rules don't explicitly allow traffic.

joestringer replied:
If you don't do this, then it's not clear in the policy trace output what is happening when a rule with only requirements is iterated over. Let's look at this in terms of the trace output. Here's what this achieves:

Search using context `From: "any:foo"` against a rule with only `FromRequires: "any:baz"`:
```
* Rule {"matchLabels":{"any:bar":""}}: selected                                 
    Enforcing requirements [{Key:any.baz Operator:In Values:[]}]                
      No label match for [any:foo]
```

Try without this logic:
```
 * Rule {"matchLabels":{"any:bar":""}}: selected
      Found all required labels
```

...and this is quite weird. If all the required labels were found, then why doesn't this rule select and allow? It doesn't explicitly say `Allows from ....`, but it doesn't say anything else either.

joestringer replied:
Perhaps I can gate this one also by the `ctx.Trace != TraceDisabled`.

ianvernon replied:
Makes sense, thanks!

ianvernon commented on pkg/policy/rule.go:
```diff
@@ -734,6 +743,12 @@ func (r *rule) resolveEgressPolicy(ctx *SearchContext, state *traceState, result
+			// provide an empty selector so that it will be updated
+			// in the next step.
+			if len(ruleCopy.ToEndpoints) == 0 {
+				ruleCopy.ToEndpoints = []api.EndpointSelector{api.NewESFromLabels()}
```
Same here.

joestringer replied:
Let's discuss in one thread, above.

ianvernon reviewed (commented):
Nothing blocking - I don't fully understand the need to create new From/ToEndpoints if rules only consist of To/FromRequires.

joestringer committed:
policy: Add SearchContext.TraceEnabled()

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Specify protocol during policy trace

This avoids failures when upcoming commits make the L4 policy trace more
strict, to determine exactly whether traffic is allowed depending on the
protocol as well as the port.

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on pkg/policy/policy.go:
```diff
@@ -106,6 +107,17 @@ func (s *SearchContext) CallDepth() string {
+func (s *SearchContext) WithLogger(log io.Writer) *SearchContext {
+	result := *s
+	result.Logging = logging.NewLogBackend(log, "", 0)
+	if result.Trace == TRACE_DISABLED {
```
Is this logic correct? We want to enable if it's disabled? Why not enable it regardless?

joestringer replied:
This was just so that if calling (test) code already has trace level at the verbose level, we won't lower the verbosity to 'enabled'.

Ie, there are 3 states and I only care to ensure that when attempting to log, it actually logs *something*.

aanm commented on pkg/policy/l4.go:
```diff
@@ -373,18 +373,21 @@ func (l4 L4PolicyMap) HasRedirect() bool {
-	if len(ports) == 0 {
-		return api.Denied
+	// Check L3-only filters first.
+	l3flt := fmt.Sprintf("0/%s", models.PortProtocolANY)
```
make `0/ANY` a const instead of creating a string every time

joestringer replied:
I'll follow up on this in another PR with the docs fixes.

joestringer replied:
https://github.com/cilium/cilium/pull/7891
